doc/decisions/1_in_discussion/copy_on_write.md

```diff
@@ -0,0 +1,474 @@
+# Copy On Write
+
+## Problem
+
+One of Elektra's core goals is low memory usage.
+Currently, there are many places within Elektra where keys and keysets are duplicated and copied around.
+Most of those copied keys are never modified.
+We want to introduce an in-memory copy-on-write mechanism to lower our memory usage.
+
+In the near future, Elektra will also gain facilities for [change tracking](../0_drafts/change_tracking.md) and session recording, both of which will potentially again duplicate keys.
+There are also aspirations to create a new, simple [internal cache](../1_in_discussion/internal_cache.md) that would also benefit from a copy-on-write mechanism.
+
+## Constraints
+
+1. Elektra's public API should require no changes
```

Comment:
```suggestion
```

Currently we do not make any API guarantees, so all changes are possible. The only restriction is that we must be able to update the whole codebase to the new API with reasonable effort. However, that is always an implied constraint, so we don't need to mention it.

A few more points regarding API changes:
1. My thesis is about migrating Elektra to a new API, so I would help with API changes
2. As part of that I'm working on a tool to automate the migrations. So if I can automate the API updates, even big API changes are possible.
3. There are already quite a few proposed API changes, including a somewhat stalled decision to change even `keyNew`. Please don't hesitate to propose something that needs big API changes. There might be a way to make it work.